NodePort with ExternalTrafficPolicy=Local should only forward traffic on nodes with local endpoints

aojea opened the pull request:
The ExternalTrafficPolicy field on services, controls if NodePort
    and LoadBalancer services send traffic to all the endpoints in the
    Service or only to the local ones.
    
    kube-proxy, with the service using ExternalTrafficPolicy=Local,
    was forwarding traffic to NodePorts with non-local endpoints.
    if the traffic was coming from pods.

/kind bug
/kind failing-test
```release-note
fixed a bug in kube-proxy that allowed pods to reach NodePort Services with ExternalTrafficPolicy=Local on nodes without local endpoints.
```

aojea commented:
we were missing coverage on this because the other tests that exercise were running under the LoadBalancer and ESIPP tags that are commonly skipped due to the need of using GCE
/assign @thockin @wojtek-t 
@swetharepakula let me know if this can help you with https://github.com/kubernetes/kubernetes/pull/99222

aojea commented:
> @swetharepakula let me know if this can help you with #99222

I just see your comment in the other PR and the problem is not this, so scratch this  comment

aojea commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1306,17 +1306,6 @@ func (proxier *Proxier) syncProxyRules() {
 					utilproxy.WriteLine(proxier.natRules, append(args, "-j", string(KubeMarkMasqChain))...)
 					// Jump to the service chain.
 					utilproxy.WriteLine(proxier.natRules, append(args, "-j", string(svcChain))...)
-				} else {
-					// TODO: Make all nodePorts jump to the firewall chain.
-					// Currently we only create it for loadbalancers (#33586).
-
-					// Fix localhost martian source error
-					loopback := "127.0.0.0/8"
-					if isIPv6 {
-						loopback = "::1/128"
-					}
-					utilproxy.WriteLine(proxier.natRules, append(args, "-s", loopback, "-j", string(KubeMarkMasqChain))...)
-					utilproxy.WriteLine(proxier.natRules, append(args, "-j", string(svcXlbChain))...)
```
@thockin this was driving me crazy because it turns out that NodePorts on nodes without local endpoints, still forward traffic if the connection comes from one pod.
I don't know if that is on purpose, can you chime in please?

aojea replied:
this was ok, I misread it

wojtek-t commented:
unassigning - I don't have enough context on that, will let networking folks to make deeper review

/unassign

Commit:
kube-proxy NodePorts and ExternalTraffic Local

The ExternalTrafficPolicy field on services, controls if NodePort
and LoadBalancer services send traffic to all the endpoints in the
Service or only to the local ones.

kube-proxy, with the service using ExternalTrafficPolicy=Local,
was forwarding traffic to NodePorts with non-local endpoints.
if the traffic was coming from pods.

aojea commented:
/assign @danwinship @andrewsykim

andrewsykim commented:
```
kube-proxy, with the service using ExternalTrafficPolicy=Local,
was forwarding traffic to NodePorts with non-local endpoints.
if the traffic was coming from pods.
```
If traffic is coming from pods, then it's not external so the policy shouldn't apply right?

andrewsykim commented:
Also, https://github.com/kubernetes/kubernetes/pull/96600 will solve this sort of thing more elegantly going forward

aojea commented:
> ```
> kube-proxy, with the service using ExternalTrafficPolicy=Local,
> was forwarding traffic to NodePorts with non-local endpoints.
> if the traffic was coming from pods.
> ```
> 
> If traffic is coming from pods, then it's not external so the policy shouldn't apply right?

I think so, but this is more than 3 years old code so it maybe a reason to work this way :shrug: 
I've found it by chance, when adding the e2e test for NodePort with externalTrafficLocal, the test failed because the pods could reach the NodePort service in ALL the nodes.

aojea commented:
> Also, #96600 will solve this sort of thing more elegantly going forward

what is holding that back?

aojea commented:
/retest

thockin reviewed (commented):
So what should we do with this now that internal policy is going alpha?

aojea commented:
> So what should we do with this now that internal policy is going alpha?

I fail to see the relation :thinking:

andrewsykim commented:
> kube-proxy, with the service using ExternalTrafficPolicy=Local,
was forwarding traffic to NodePorts with non-local endpoints.
if the traffic was coming from pods.

If traffic is coming from pods, then `externalTrafficPolicy` doesn't apply, because traffic is "internal"

We never had a knob to change this behavior for internal traffic, but with https://github.com/kubernetes/kubernetes/pull/96600 close to merging we do now. So if a user wants to changing this specific behavior for traffic from pods, they can use `internalTrafficPolicy: Local`

danwinship commented:
yeah I think this is NOTABUG?

aojea commented:
ah so the pod -> nodePort is not external, got it now, fair enough
